HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+from HARK import MetricObject, NullFunc  # Basic HARK features
+from HARK.ConsumptionSaving.ConsIndShockModel import (
+    utility,  # CRRA utility function
+    utility_inv,  # Inverse CRRA utility function
+    utilityP,  # CRRA marginal utility function
+    utilityP_inv,  # Inverse CRRA marginal utility function
+    init_lifecycle,
+)
+
+from HARK.ConsumptionSaving.ConsRiskyAssetModel import (
+    RiskyAssetConsumerType,
+    risky_asset_parms,
+    init_risky_asset,
+)
+
+from HARK.distribution import calc_expectation
+
+from HARK.interpolation import (
+    LinearInterp,  # Piecewise linear interpolation
+    BilinearInterp,  # 2D interpolator
+    TrilinearInterp,  # 3D interpolator
+    ConstantFunction,  # Interpolator-like class that returns constant value
+    IdentityFunction,  # Interpolator-like class that returns one of its arguments
+    ValueFuncCRRA,
+    MargValueFuncCRRA,
+    DiscreteInterp,
+)
+
+from HARK.utilities import make_grid_exp_mult
+
+
+class RiskyContribConsumerType(RiskyAssetConsumerType):
+    """
+    A consumer type with idiosyncratic shocks to permanent and transitory income,
+    who can save in both a risk-free and a risky asset but faces frictions to
+    moving funds between them. The agent can only consume out of his risk-free
+    asset.
+    
+    The frictions are:
+        - A proportional tax on funds moved from the risky to the risk-free
+         asset.
+        - A stochastic inability to move funds between his accounts.
+    
+    To partially avoid the second friction, the agent can commit to have a
+    fraction of his labor income, which is usually deposited in his risk-free
+    account, diverted to his risky account. He can change this fraction
+    only in periods where he is able to move funds between accounts.
+    """
+
+    time_inv_ = deepcopy(RiskyAssetConsumerType.time_inv_)
+    time_inv_ = time_inv_ + ["DiscreteShareBool"]
+
+    state_vars = RiskyAssetConsumerType.state_vars + [
+        "nNrm",
+        "mNrmTilde",
+        "nNrmTilde",
+        "Share",
+    ]
+    shock_vars_ = RiskyAssetConsumerType.shock_vars_
+
+    def __init__(self, cycles=1, verbose=False, quiet=False, **kwds):
+
+        params = init_risky_contrib.copy()
+        params.update(kwds)
+        kwds = params
+
+        # Initialize a basic consumer type
+        RiskyAssetConsumerType.__init__(
+            self, cycles=cycles, verbose=verbose, quiet=quiet, **kwds
+        )
+
+        # The model is solved and simulated spliting each of the agent's
+        # decisions into its own "stage". The stages in chronological order
+        # are
+        # - Reb: asset-rebalancing stage.
+        # - Sha: definition of the income contribution share.
+        # - Cns: consumption stage.
+        self.stages = ["Reb", "Sha", "Cns"]
+
+        # Each stage has its own states and controls, and its methods
+        # to find them.
+        self.set_states = {
+            "Reb": self.get_states_Reb,
```

Comment:
This construction is a little awkward. "getter" and "setter" have established meanings in software engineering as a coding pattern -- they are ways to access a member attribute of an object.

If I'm not mistaken, you are here defining `set_states` to be a dictionary of methods that then "get" (maybe, assign, or compute) states.

Have you considered other data structures that might allow for more modularity?

For example, what if `Stage` was a class, with states and controls methods?

Reply:
While I agree with you---these functions are computing and "setting" states---the terminology I settled on came from [core's sim_one_period()](https://github.com/econ-ark/HARK/blob/d5ca85009c88e6fded02c684b30682288370bfe9/HARK/core.py#L556) which uses `get_states()`, `get_controls()`, `get_poststates()`. I guess the conflicting terminology of "get" methods inside a "set" dictionary came from these conflicting forces. I'll just make them all "get" for consistency with the existing parts of HARK.

What I liked about a dictionary was being able to call them by name. I agree that a better data structure might exist, but I'd like to keep this provisional solution while a consensus is reached an implemented on how to deal with "stages", "frames", etc.

Reply:
Addressed in https://github.com/econ-ark/HARK/pull/832/commits/f7915186107e9905487ceef8d23c7f2f4364e5ea